### PR TITLE
Fix/sof es8336 quirk

### PR DESCRIPTION
**Pull request by yangxiaohua2009**
The reason why ES8326 codec driver cannot be probed is that there is something wrong with the ACPI so that every module related to it can not be probed, for example the i2c_hid_acpi module. The i2c bus driver is successfully probed though.

Anyway this should be a harmless update which adds support for ES8326 codec and can be tested by the users.

ASoC: codecs: add support for ES8326

The ES8326 codec is not compatible with ES8316 and requires a dedicated driver.

Signed-off-by: David Yang <yangxiaohua@everest-semi.com>

### CODE REVIEW

**Review by plbossart — CHANGES_REQUESTED** [review]
I can't even being to review this unless I see a Signed-off-by tag.

How difficult is this to add this to the commit message:

````
ASoC: codecs: add support for ES8326

The ES8326 codec is not compatible with ES8316 and requires a dedicated driver.

Signed-off-by: David Yang <yangxiaohua@everest-semi.com>
````

**Comment by yangxiaohua2009** [comment]
edited

File: sound/soc/codecs/es8326.c [hunk 1]
```diff
@@ -0,0 +1,854 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * es8326.c -- es8326 ALSA SoC audio driver
+ * Copyright Everest Semiconductor Co.,Ltd
+ *
+ * Authors: David Yang <yangxiaohua@everest-semi.com>
+ */
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:7]: Mark Brown will ask you to use C++ // comments for all the lines

File: sound/soc/codecs/es8326.c [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-dapm.h>
+#include <sound/tlv.h>
+#include "es8326.h"
+
+#define ES8326_LRCK 48000
+
+static int es8326_set_bias_level(struct snd_soc_component*codec,
+								enum snd_soc_bias_level level);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: alignment

File: sound/soc/codecs/es8326.c [hunk 3]
```diff
@@ -0,0 +1,854 @@
...
+#include <linux/irq.h>
+#include <linux/module.h>
+#include <sound/jack.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-dapm.h>
+#include <sound/tlv.h>
+#include "es8326.h"
+
+#define ES8326_LRCK 48000
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: This should come from hw_params

File: sound/soc/codecs/es8326.c [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+		{0x40, 0xff}, {0x41, 0x1f}, {0x42, 0xf7}, {0x43, 0xfd},
+		{0x44, 0xff}, {0x45, 0x1f}, {0x46, 0xf7}, {0x47, 0xfd},
+		{0x48, 0xff}, {0x49, 0x1f}, {0x4a, 0xf7}, {0x4b, 0xfd},
+		{0x4c, 0xff}, {0x4d, 0x00}, {0x4e, 0x00}, {0x4f, 0xff},
+		{0x50, 0x00}, {0x51, 0x00}, {0x52, 0x00}, {0x53, 0x00},
+		{0x54, 0x00}, {0x55, 0x00}, {0x56, 0x00}, {0x57, 0x1f},
+		{0x58, 0x00}, {0x59, 0x00}, {0x5a, 0x00}, {0x5b, 0x00},
+		{0x5c, 0x00}, {0xf9, 0x00}, {0xfa, 0x00}, {0xfb, 0x00},
+		{0xfc, 0x00}, {0xfd, 0x00}, {0xfe, 0x00}, {0xff, 0x00},
+};
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:54]: I think your editor is off, for kernel style the indentation is 8 characters per tab.

File: sound/soc/codecs/es8326.c [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+		{2048,8000,16384000,0x60,0x00,0x00,0x35,0x0A,0x1B,0x1F,0x7F}, 
+		{2304,8000,18432000,0x40,0x02,0x10,0x35,0x0A,0x1B,0x1F,0x5F}, 
+		{3072,8000,24576000,0x60,0x02,0x10,0x35,0x0A,0x1B,0x1F,0x7F}, 
+		{3250,8000,26000000,0x0C,0x18,0x0F,0x2D,0x0A,0x0A,0x27,0x27}, 
+
+};
+
+static inline int get_coeff(int mclk, int rate)
+{
+		int i;
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: newline after this

File: sound/soc/codecs/es8326.c [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+		rate=ES8326_LRCK;
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: why would you override the rate instead of using the parameter?

File: sound/soc/codecs/es8326.c [hunk 7]
```diff
@@ -0,0 +1,854 @@
...
+		for (i = 0; i < ARRAY_SIZE(coeff_div); i++) {
+				if (coeff_div[i].rate == rate && coeff_div[i].mclk == mclk)
+						return i;
+		}
+
+		return -EINVAL;
+}
+
+static unsigned int rates_12288[] = {
+		8000, 12000, 16000, 24000, 24000, 32000, 48000, 96000,
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: repeated 24000?

File: sound/soc/codecs/es8326.c [hunk 8]
```diff
@@ -0,0 +1,854 @@
...
+};
+
+static struct snd_pcm_hw_constraint_list constraints_112896 = {
+		.count = ARRAY_SIZE(rates_112896),
+		.list = rates_112896,
+};
+
+static unsigned int rates_12[] = {
+		8000, 11025, 12000, 16000, 22050, 24000, 32000, 44100, 48000,
+		48000, 88235, 96000,
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: 88200?

I have never seen 88235 ever being used?

File: sound/soc/codecs/es8326.c [hunk 9]
```diff
@@ -0,0 +1,854 @@
...
+}
+
+static int es8326_set_dai_fmt(struct snd_soc_dai *codec_dai, unsigned int fmt)
+{
+	
+		struct snd_soc_component*component = codec_dai->component;
+		u8 iface = 0;
+
+		switch (fmt & SND_SOC_DAIFMT_MASTER_MASK) {
+		case SND_SOC_DAIFMT_CBM_CFM:
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: CBP_CFP

File: sound/soc/codecs/es8326.c [hunk 10]
```diff
@@ -0,0 +1,854 @@
...
+		return -EINVAL;	
+}
+
+static int es8326_set_dai_fmt(struct snd_soc_dai *codec_dai, unsigned int fmt)
+{
+	
+		struct snd_soc_component*component = codec_dai->component;
+		u8 iface = 0;
+
+		switch (fmt & SND_SOC_DAIFMT_MASTER_MASK) {
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: SND_SOC_DAIFMT_CLOCK_PROVIDER_MASK

File: sound/soc/codecs/es8326.c [hunk 11]
```diff
@@ -0,0 +1,854 @@
...
+		case SND_SOC_DAIFMT_CBM_CFM:	
+				snd_soc_component_update_bits(component, ES8326_RESET_00, 
+							ES8326_MASTER_MODE_EN, ES8326_MASTER_MODE_EN);
+				break;
+		case SND_SOC_DAIFMT_CBS_CFS:
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: CBC_CFC

File: sound/soc/codecs/es8326.c [hunk 12]
```diff
@@ -0,0 +1,854 @@
...
+		default:
+				return -EINVAL;
+		}
+
+		/* interface format */
+		switch (fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
+		case SND_SOC_DAIFMT_I2S:
+				break;
+		case SND_SOC_DAIFMT_RIGHT_J:
+				dev_err(component->dev, "Codec driver don't support right justified\n");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: does not

File: sound/soc/codecs/es8326.c [hunk 13]
```diff
@@ -0,0 +1,854 @@
...
+				break;
+		case SNDRV_PCM_FORMAT_S18_3LE:
+				srate |= ES8326_S18_LE;
+				break;
+		case SNDRV_PCM_FORMAT_S24_LE:
+				srate |= ES8326_S24_LE;
+				break;
+		case SNDRV_PCM_FORMAT_S32_LE:
+				srate |= ES8326_S32_LE;
+				break;
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: default case is needed

File: sound/soc/codecs/es8326.c [hunk 14]
```diff
@@ -0,0 +1,854 @@
...
+
+
+
+static int es8326_set_bias_level(struct snd_soc_component*codec,
+				 enum snd_soc_bias_level level)
+{
+		struct es8326_priv *es8326 = snd_soc_component_get_drvdata(codec);
+		int ret;
+
+		dev_dbg(codec->dev, "Entering %s\n",__func__);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: remove

File: sound/soc/codecs/es8326.c [hunk 15]
```diff
@@ -0,0 +1,854 @@
...
+		switch (level) {
+		case SND_SOC_BIAS_ON:
+				break;
+		case SND_SOC_BIAS_PREPARE:
+				dev_dbg(codec->dev, "%s prepare\n", __func__);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: remove

File: sound/soc/codecs/es8326.c [hunk 16]
```diff
@@ -0,0 +1,854 @@
...
+		.ops = &es8326_ops,
+		.symmetric_rates = 1,
+};
+
+static int es8326_probe(struct snd_soc_component*codec)
+{
+		struct es8326_priv *es8326 = snd_soc_component_get_drvdata(codec);
+		es8326->component = codec;
+
+		if (codec == NULL) {
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: Can this happen?

Also even if it's needed, you have an NULL dereference on line 606

File: sound/soc/codecs/es8326.c [hunk 17]
```diff
@@ -0,0 +1,854 @@
...
+		snd_soc_component_write(codec, ES8326_DAC_DSM_4D, 0x08);
+		snd_soc_component_write(codec, ES8326_DAC_RAMPRATE_4E, 0x20);
+		snd_soc_component_write(codec, ES8326_DAC_VPPSCALE_4F, 0x15);
+		snd_soc_component_write(codec, ES8326_HPJACK_TIMER_56, 0x88);
+		snd_soc_component_write(codec, ES8326_HP_DET_57, ES8326_HP_DET_SRC_PIN9|es8326->jack_pol);
+		snd_soc_component_write(codec, ES8326_INT_SOURCE_58, 0x08);
+		snd_soc_component_write(codec, ES8326_INTOUT_IO_59, 0x45);
+		snd_soc_component_write(codec, ES8326_RESET_00, ES8326_CSM_ON);
+		snd_soc_component_update_bits(codec, ES8326_PAGGAIN_23, 
+				ES8326_MIC_SEL_MASK, ES8326_MIC1_SEL);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: is there a reason you don't use regmap for those?

File: sound/soc/codecs/es8326.c [hunk 18]
```diff
@@ -0,0 +1,854 @@
...
+		struct es8326_priv *es8326;
+		u8 reg;
+		int ret = -1;
+		struct i2c_adapter *adapter = to_i2c_adapter(i2c->dev.parent);
+
+		if (!i2c_check_functionality(adapter, I2C_FUNC_I2C)) {
+				dev_warn(&adapter->dev,
+					"I2C-Adapter doesn't support I2C_FUNC_I2C\n");
+				return -EIO;
+		}
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: I don't think this block is needed, it's not in any existing codec driver.

File: sound/soc/codecs/es8326.c [hunk 19]
```diff
@@ -0,0 +1,854 @@
...
+
+		es8326 = devm_kzalloc(&i2c->dev, sizeof(struct es8326_priv), GFP_KERNEL);
+		if (es8326 == NULL)
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: if (!es8326)

run scripts/checkpatch.pl --strict --codespell on your patches please.

File: sound/soc/codecs/es8326.c [hunk 20]
```diff
@@ -0,0 +1,854 @@
...
+				dev_err(&i2c->dev,"%s, unable to enable mclk\n", __func__);
+				return ret;
+		}
+
+		reg = 0x00;
+		ret = i2c_master_recv(i2c, &reg, 1);
+		if (ret < 0) {
+				dev_err(&i2c->dev, "i2c recv Failed\n");
+				return ret;
+		}
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: that seems also very odd.

File: sound/soc/codecs/es8326.c [hunk 21]
```diff
@@ -0,0 +1,854 @@
...
+		es8326->irq = i2c->irq;
+		ret = devm_request_threaded_irq(&i2c->dev, es8326->irq, NULL, es8326_irq,
+									IRQF_TRIGGER_RISING | IRQF_ONESHOT,
+									"es8326", es8326);
+		if(ret == 0)	
+		{
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: braces at end of line for kernel style

File: sound/soc/codecs/es8326.c [hunk 22]
```diff
@@ -0,0 +1,854 @@
...
+				es8326->jack_pol = ES8326_HP_DET_JACK_POL|ES8326_HP_DET_BUTTON_POL|ES8326_HP_TYPE_OMTP;
+		}
+		dev_dbg(&i2c->dev, "jack-pol %x",es8326->jack_pol);
+
+		ret = of_property_read_u32(i2c->dev.of_node, "mclk-rate", &es8326->mclk_rate);
+		if (ret != 0) {
+				dev_dbg(&i2c->dev, "mclk-rate return %d", ret);
+				es8326->mclk_rate = 19200000;
+		}
+		dev_dbg(&i2c->dev, "mclk-rate %u",es8326->mclk_rate);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: I don't think this can work, we need to read the properties in the component probe if you want the machine driver to modify them. This is how it works for es8316.c

In addition, please use device_property instead of of_property so that it also works for ACPI platforms.

Comment (plbossart) [inline sound/soc/codecs/es8326.c]: what about jack detect inversion, is this not supported?

Comment (plbossart) [inline sound/soc/codecs/es8326.c]: you will also need DT bindings for all these properties.

File: sound/soc/codecs/es8326.h [hunk 23]
```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
Comment (plbossart) [inline sound/soc/codecs/es8326.h]: here you need C comments :-)

File: sound/soc/codecs/es8326.c [hunk 24]
```diff
@@ -0,0 +1,854 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:1]: The commit title should be
ASoC: Add support for ES8326 codec

**Review by plbossart — CHANGES_REQUESTED** [review]
This looks like a good start. You have a lot of style issues, please run checkpatch and fix the indentation/style issues. The more important problem is how properties are handled, they need to be read in the component probe so that the machine driver can change the values of the defaults.

File: sound/soc/codecs/es8326.c (hunk 2 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: the tab size is now 8 instead of 4

File: sound/soc/codecs/es8326.c (hunk 6 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: because I cann't get the parameter. It's defined in tplg file. If the user use `arecord 1.wav -r 8000`, the LRCK rate will still be 48000.

File: sound/soc/codecs/es8326.c [hunk 25]
```diff
@@ -0,0 +1,854 @@
...
+
+static struct snd_pcm_hw_constraint_list constraints_12 = {
+		.count = ARRAY_SIZE(rates_12),
+		.list = rates_12,
+};
+
+/*
+ * Note that this should be called from init rather than from hw_params.
+ */
+static int es8326_set_dai_sysclk(struct snd_soc_dai *codec_dai,
```
Comment (yangxiaohua2009) [inline sound/soc/codecs/es8326.c:362]: Can I get rid of the constraints?  I tried to remove them but then I see a kernel panic

File: sound/soc/codecs/es8326.c (hunk 17 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: Is there any difference?

File: sound/soc/codecs/es8326.c (hunk 19 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: checked.

File: sound/soc/codecs/es8326.c (hunk 22 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: defined as ES8326_HP_DET_JACK_POL. Is a device properity named jd_inverted necessary?

File: sound/soc/codecs/es8326.c (hunk 1 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c:7]: I followed the style of other codecs drivers. Is it ok?

File: sound/soc/codecs/es8326.c (hunk 22 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c]: These properities are related to hp detection and mic source. The dapm can handle them, but a custom UCM would be needed.

File: sound/soc/codecs/es8326.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c:7]: let's see.

File: sound/soc/codecs/es8326.c (hunk 17 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c]: regmap reduces the work needed for e.g. suspend-resume and only accesses registers that need to be modified.

As a minimum, you will need to explain to Mark Brown with a comment why you chose direct register access.

File: sound/soc/codecs/es8326.c (hunk 22 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c]: well, if you want to use this codec driver with the same machine driver, then the device properties and quirks have to be supported for all codec drivers. I don't want to special case this es8326 driver for property handling.

Comment (plbossart) [reply sound/soc/codecs/es8326.c]: I don't think you got my point. a DT binding is REQUIRED for upstream. not my decision, it's just the way things are.

If you don't provide this DT binding then this code will NEVER land upstream.

File: sound/soc/codecs/es8326.c (hunk 25 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c:362]: that's your code, I don't know what this does.

**Comment by plbossart** [comment]
@yangxiaohua2009 please squash patches and do a git push --force to update this PR. I don't want to review white space changes.

File: sound/soc/codecs/es8326.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c]: sure you do, you get the hw_params and can get the rate defined by topology. the LRCK is the samplerate or frame rate, it's propagated to all components by ASoC.

**Comment by plbossart** [comment]
Sorry, @yangxiaohua2009 you need to clean-up your code before asking for reviews. I am not being painful or setting the bar high, this is just basic stuff.

scripts/checkpatch.pl --strict --codespell *.patch &> checkpatch.log
total: 708 errors, 92 warnings, 64 checks, 1038 lines checked

[checkpatch.log](https://github.com/plbossart/sound/files/8128934/checkpatch.log)

**Comment by yangxiaohua2009** [comment]
> Sorry, @yangxiaohua2009 you need to clean-up your code before asking for reviews. I am not being painful or setting the bar high, this is just basic stuff.
> 
> scripts/checkpatch.pl --strict --codespell *.patch &> checkpatch.log total: 708 errors, 92 warnings, 64 checks, 1038 lines checked
> 
> [checkpatch.log](https://github.com/plbossart/sound/files/8128934/checkpatch.log)

already cleaned

**Comment by plbossart** [comment]
> > Sorry, @yangxiaohua2009 you need to clean-up your code before asking for reviews. I am not being painful or setting the bar high, this is just basic stuff.
> > scripts/checkpatch.pl --strict --codespell *.patch &> checkpatch.log total: 708 errors, 92 warnings, 64 checks, 1038 lines checked
> > [checkpatch.log](https://github.com/plbossart/sound/files/8128934/checkpatch.log)
> 
> already cleaned

come on, man. Even when I squash your patches, you still have quite a few issues in there - the worst being no Signed-off-by tags.

````
WARNING: Missing commit description - Add an appropriate one

WARNING: please write a paragraph that describes the config symbol fully
#31: FILE: sound/soc/codecs/Kconfig:867:
+config SND_SOC_ES8326

ERROR: do not set execute permissions for source files
#59: FILE: sound/soc/codecs/es8326.c

CHECK: Alignment should match open parenthesis
#88: FILE: sound/soc/codecs/es8326.c:25:
+static int es8326_set_bias_level(struct snd_soc_component *codec,
+				enum snd_soc_bias_level level);

CHECK: struct mutex definition without comment
#135: FILE: sound/soc/codecs/es8326.c:72:
+	struct mutex lock;


CHECK: Blank lines aren't necessary after an open brace '{'
#516: FILE: sound/soc/codecs/es8326.c:453:
+{
+

CHECK: Alignment should match open parenthesis
#567: FILE: sound/soc/codecs/es8326.c:504:
+static int es8326_set_bias_level(struct snd_soc_component *codec,
+				enum snd_soc_bias_level level)

CHECK: Alignment should match open parenthesis
#742: FILE: sound/soc/codecs/es8326.c:679:
+static void es8326_enable_jack_detect(struct snd_soc_component *component,
+				struct snd_soc_jack *jack)

CHECK: Alignment should match open parenthesis
#778: FILE: sound/soc/codecs/es8326.c:715:
+static int es8326_set_jack(struct snd_soc_component *component,
+			struct snd_soc_jack *jack, void *data)

ERROR: do not set execute permissions for source files
#887: FILE: sound/soc/codecs/es8326.h

WARNING: Improper SPDX comment style for 'sound/soc/codecs/es8326.h', please use '/*' instead
#892: FILE: sound/soc/codecs/es8326.h:1:
+// SPDX-License-Identifier: GPL-2.0-only

WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#892: FILE: sound/soc/codecs/es8326.h:1:
+// SPDX-License-Identifier: GPL-2.0-only

ERROR: trailing whitespace
#1005: FILE: sound/soc/codecs/es8326.h:114:
+#define ES8326_S16_LE^I(3 << 2)^I^I$

CHECK: spaces preferred around that '|' (ctx:VxV)
#1009: FILE: sound/soc/codecs/es8326.h:118:
+#define ES8326_DAIFMT_MASK	((1 << 5)|(3 << 0))
                           	         ^
ERROR: trailing whitespace
#1012: FILE: sound/soc/codecs/es8326.h:121:
+#define ES8326_DAIFMT_DSP_A^I^I(3 << 0)^I^I$

CHECK: spaces preferred around that '|' (ctx:VxV)
#1013: FILE: sound/soc/codecs/es8326.h:122:
+#define ES8326_DAIFMT_DSP_B		((1 << 5)|(3 << 0))
                            		         ^
CHECK: spaces preferred around that '|' (ctx:ExV)
#1036: FILE: sound/soc/codecs/es8326.h:145:
+		|(ES8326_ADC_SRC_ANA_INV_SW1 << ES8326_ADC1_SHIFT))
 		^

CHECK: spaces preferred around that '|' (ctx:ExV)
#1038: FILE: sound/soc/codecs/es8326.h:147:
+		|(ES8326_ADC_SRC_DMIC_SDIN2 << ES8326_ADC1_SHIFT))
 		^

ERROR: Missing Signed-off-by: line(s)

````

**Comment by yangxiaohua2009** [comment]
devicetree bindings added. Didn't see properity "everest,jack-detect-inverted" in `Documentation/devicetree/bindings/sound/everest,es8316.yaml`
Is it defined elsewhere?

**Comment by plbossart** [comment]
> devicetree bindings added. Didn't see properity "everest,jack-detect-inverted" in `Documentation/devicetree/bindings/sound/everest,es8316.yaml` Is it defined elsewhere?

it's a miss. Not sure how this went undetected during reviews. This needs to be fixed with a correction reflecting all the properties supported by this driver.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+
+  amic1-src:
+    description:
+      the value of reg 23 when headset plugged.
+    $ref: /schemas/types.yaml#/definitions/uint8-array
+    minimum: 0x19
+    maximum: 0x29
+    default: 0x19
+
+  amic2-src:
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml:71]: don't you need the 'everest,' prefix for all properties?

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+
+required:
+  - compatible
+  - reg
+  - "#sound-dai-cells"
+  - mclk-rate
+  - mic1-src
+  - mic2-src
+  - amic1-src
+  - amic2-src
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: they are certainly not required since we don't use them from the machine driver today...

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+        reg = <0x19>;
+        clocks = <&clks 10>;
+        clock-names = "mclk";
+        #sound-dai-cells = <0>;
+        mic1-src = [22];
+        mic2-src = [44];
+        amic1-src = [19];
+        amic2-src = [19];
+        jack-pol = [0e];
+        mclk-rate = <12288000>;
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml]: the mclk rate is typically set with snd_soc_dai_set_sysclk(), you don't need to make this a property.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+
+  mclk-rate:
+    description:
+      clock rate of mclk
+    $ref: /schemas/types.yaml#/definitions/uint32-array
+    minItems: 1
+
+  jack-pol:
+    description:
+      just the value of reg 57. Bit(3) decides whether the jack polarity is inverted.
```
Comment (plbossart) [inline Documentation/devicetree/bindings/sound/everest,es8326.yaml:39]: how does this work since you also have the 'jack-inverted' value?

File: sound/soc/codecs/es8326.c [hunk 30]
```diff
@@ -0,0 +1,819 @@
...
+	struct snd_soc_component *component;
+	bool amic;
+	bool start;
+	bool muted;
+	bool hp_inserted;
+	bool spk_gpio_level;
+	bool hp_det_level;
+	struct snd_soc_jack *jack;
+	int irq;
+	struct mutex lock;
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:70]: add a comment on what this protects

File: sound/soc/codecs/es8326.c [hunk 31]
```diff
@@ -0,0 +1,819 @@
...
+{
+	struct es8326_priv *es8326 = dev_id;
+	struct snd_soc_component *comp = es8326->component;
+	unsigned int iface;
+
+	mutex_lock(&es8326->lock);
+	if (!es8326->jack)
+		goto out;
+	snd_soc_component_write(comp, ES8326_ANA_MICBIAS_1B, 0x7c);
+	mdelay(500);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:90]: add a newline after this, it's hard to read otherwise

File: sound/soc/codecs/es8326.c [hunk 32]
```diff
@@ -0,0 +1,819 @@
...
+static int es8326_pcm_hw_params(struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params,
+				struct snd_soc_dai *dai)
+{
+	struct snd_soc_component *codec = dai->component;
+	struct es8326_priv *es8326 = snd_soc_component_get_drvdata(codec);
+	u8 srate = 0;
+	int coeff;
+
+	coeff = get_coeff(es8326->mclk_rate, params_rate(params));
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: I don't understand why you need the clock to be handled with a codec property. See es8316, this is handled with the clk_ API, 

````c
	es8316->mclk = devm_clk_get_optional(component->dev, "mclk");
	if (IS_ERR(es8316->mclk)) {
		dev_err(component->dev, "unable to get mclk\n");
		return PTR_ERR(es8316->mclk);
	}
	if (!es8316->mclk)
		dev_warn(component->dev, "assuming static mclk\n");

	ret = clk_prepare_enable(es8316->mclk);
	if (ret) {
		dev_err(component->dev, "unable to enable mclk\n");
		return ret;
	}
````

File: sound/soc/codecs/es8326.c [hunk 33]
```diff
@@ -0,0 +1,819 @@
...
+
+	switch (level) {
+	case SND_SOC_BIAS_ON:
+		break;
+	case SND_SOC_BIAS_PREPARE:
+		if (!IS_ERR(es8326->mclk)) {
+			if (snd_soc_component_get_bias_level(codec) == SND_SOC_BIAS_ON) {
+				clk_disable_unprepare(es8326->mclk);
+			} else {
+				ret = clk_prepare_enable(es8326->mclk);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:514]: and here you do use the clk API, so why did you need a property at all?

File: sound/soc/codecs/es8326.c [hunk 34]
```diff
@@ -0,0 +1,819 @@
...
+		es8326->jack_pol = ES8326_HP_DET_BUTTON_POL | ES8326_HP_TYPE_OMTP;
+	}
+	dev_dbg(codec->dev, "jack-pol %x", es8326->jack_pol);
+
+	ret = device_property_read_u32(codec->dev, "mclk-rate", &es8326->mclk_rate);
+	if (ret != 0) {
+		dev_dbg(codec->dev, "mclk-rate return %d", ret);
+		es8326->mclk_rate = 19200000;
+	}
+	dev_dbg(codec->dev, "mclk-rate %u", es8326->mclk_rate);
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: the mclk rate should not be a property.

File: sound/soc/codecs/es8326.c [hunk 35]
```diff
@@ -0,0 +1,819 @@
...
+		dev_err(&i2c->dev, "%s, unable to get mclk\n", __func__);
+		return PTR_ERR(es8326->mclk);
+	}
+	if (!es8326->mclk)
+		dev_err(&i2c->dev, "%s, assuming static mclk\n", __func__);
+		ret = clk_prepare_enable(es8326->mclk);
+	if (ret) {
+		dev_err(&i2c->dev, "%s, unable to enable mclk\n", __func__);
+		return ret;
+	}
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:765]: what prevents you from using the same code as in es8316.c? the code is virtually identical, so why add variations?

File: sound/soc/codecs/es8326.c [hunk 36]
```diff
@@ -0,0 +1,819 @@
...
+
+	es8326->irq = i2c->irq;
+	ret = devm_request_threaded_irq(&i2c->dev, es8326->irq, NULL, es8326_irq,
+					IRQF_TRIGGER_RISING | IRQF_ONESHOT,
+					"es8326", es8326);
+	if (ret == 0)
+		disable_irq(es8326->irq);
+	else
+		dev_warn(&i2c->dev, "Getting irq failed.");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c:774]: again why not use the exact same sequence as es8316.c, and explain why the IRQF_ parameters are different?

File: sound/soc/codecs/es8326.c [hunk 37]
```diff
@@ -0,0 +1,819 @@
...
+		.acpi_match_table = ACPI_PTR(es8326_acpi_match),
+		.of_match_table = of_match_ptr(es8326_of_match),
+	},
+	.probe = es8326_i2c_probe,
+};
+module_i2c_driver(es8326_i2c_driver);
+
+MODULE_DESCRIPTION("ASoC es8326 driver");
+MODULE_AUTHOR("David Yang <yangxiaohua@everest-semi.com>");
+MODULE_LICENSE("GPL v2");
```
Comment (plbossart) [inline sound/soc/codecs/es8326.c]: GPL

**Review by plbossart — CHANGES_REQUESTED** [review]
It's much better @yangxiaohua2009 but the clk handling needs to be revisited and you need to align with es8316.c when possible, and explain differences if required. This will be very useful for maintainers.

File: sound/soc/codecs/es8326.c (hunk 36 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c:774]: IRQF_  difference is hardware issue. es8316 generates a high voltage as irq while es8326 generates an impulse.

File: sound/soc/codecs/es8326.c (hunk 30 above)
Comment (yangxiaohua2009) [reply sound/soc/codecs/es8326.c:70]: copyed from es8316.c.
The lock protects the situation that an irq is generated while the previous irq is still being processed.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 29 above)
Comment (yangxiaohua2009) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml:39]: A default value is defined here which can be inverted by 'jack-inverted' ？

File: sound/soc/codecs/es8326.c (hunk 36 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c:774]: add a comment then that the es8316 is level-based while es8326 is edge-based.

File: sound/soc/codecs/es8326.c (hunk 30 above)
Comment (plbossart) [reply sound/soc/codecs/es8326.c:70]: you want to add this comment in the code itself.

File: Documentation/devicetree/bindings/sound/everest,es8326.yaml (hunk 29 above)
Comment (plbossart) [reply Documentation/devicetree/bindings/sound/everest,es8326.yaml:39]: ok

**Comment by plbossart** [comment]
@yangxiaohua2009 please resubmit this as a PR against the SOF tree https://github.com/thesofproject/linux branch topic/sof-dev, thanks!
